CustomDialogOnClose/explainer.md

```diff
@@ -0,0 +1,125 @@
+   - There should be a time limit on how long the asynchronous code can run for, at the end of the time limit there should be a "leave anyway" dialog, or some other mechanism that allows the user to cancel the async task and leave immediately.
+   - A second attempt to close the tab, via clicking the "x" or ctrl+w, should immediately close the tab.
+ - A website displays alarming text, to try and convince the user to stay on the site or perform some harmful action.
+   - This capability should only be available to installed PWAs, and not arbitrary websites.
+   - The text that the website can display should be limited in length and not allow formatting.
+   - Always display "Leave" and "Cancel" buttons on the dialog, and only allow websites to add a third option which could be customizable.
+
+## Considered Alternatives
+### Extend window.confirm()
+[`window.confirm()`](https://developer.mozilla.org/en-US/docs/Web/API/Window/confirm) is a familiar API that lets websites query the user and act based on their response.
+
+![A dialog is displayed with the title "An embedded page on this page says" and the contents "This is an example of the confirm API, this text is customizable". There are two buttons "OK" and "Cancel".](./images/ConfirmExample.PNG)
+
+We could extend this API to allow websites to add a new options to the dialog and have it return a promise with the user's selection. The website would then return a promise by calling .then() on window.confirm(). When the returned promise settles, the User Agent will close the tab. For example:
+    
+For example:
+```javascript
+window.addEventListener("beforeunload", (event) => {
+  if (featureDetectNewConfirmDialog()) {
+    return window.confirm("You have unsaved changes, would you like to save them?", "Save & Leave").then(function(result) {
+      // Perform the work to save the document.
+      resolve(); // The tab closes at this point.
+    });
+  } else {
+    event.preventDefault();
+  }
+});
+```
+
+However, this proposal brings confusion because window.confirm() would behave differently depending on if it was called from a beforeunload handler or not. It is also not a common pattern, and currently User Agents are allowed to deny showing window.confirm() dialogs from the beforeunload handler. For these reasons, this proposal was not chosen.
+
+### Fully customizable dialog
+
+We considered allowing the developer to design and show a fully customizable dialog, perhaps via the proposed [modal-window](https://github.com/adrianhopebailie/modal-window/blob/master/explainer.md), but this was not chosen due to it's additional complexity and the increased potential for abuse. For example, the site could choose to show only an "Accept" option, and no option to cancel. While mitigations for problems like this exist, the user experience would be worse than the proposed solution, where it is guaranteed that the user is given the option to leave a site immediately.
+
+## Privacy
```

Comment:
How about simplifying this paragraph to something like, "This extension of the `beforeunload` event would increase browser fingerprinting capabilities by adding a feature-detectable API. However, as this API holds no user preferences, fingerprinting capabilities are limited to detecting large cohorts of users of particular browsers."

Reply:
Perhaps a better way to write that, but explain why this fingerprinting capability would be of minimal impact :)

Reply:
Good suggestion, I've updated it with the text you provided.